File: src/sorting/heapSort.cs
using System;
using Sorting;

namespace HeapSort
{
    public class HeapSort<T> : Sorting<T> where T : IComparable<T>
    {
        public HeapSort()
        {
            _arr = Array.ConvertAll<string, T>(Console.ReadLine().Split(' ', 
            StringSplitOptions.RemoveEmptyEntries), ParseType);
            _count = _arr.Length;
        }

        public HeapSort(T[] input)
        {
            _arr = input;
            _count = input.Length;
        }

        private void Heapify(int size, int tIdx)
        {
            int largest = tIdx;
            int left = tIdx * 2 + 1;
            int right = tIdx * 2 + 2;

            if (left < size && _arr[left].CompareTo(_arr[largest]) > 0)
                largest = left;
            if (right < size && _arr[right].CompareTo(_arr[largest]) > 0)
                largest = right;

            if (largest != tIdx)
            {
                Swap(largest, tIdx);
                Heapify(size, largest);
            }
            return;
        }

        public override void Sort()
        {
            for (int i = _count / 2 - 1; i >= 0; i--)
                Heapify(_count, i);
            for (int i = _count - 1; i >= 0; i--)
            {
                Swap(0, i);
                Heapify(i, 0);
            }
        }
    }

    public static class Program
    {
        static void Main()
        {
            Console.OutputEncoding = System.Text.Encoding.UTF8;
            var n = new HeapSort<int>();
            n.PrintArray();
            n.Sort();
            n.PrintArray();
        }
    }
}

/*
Input:
5 4 3 2 1

Output:
5 4 3 2 1
1 2 3 4 5
*/

File: src/sorting/heapSort.py
def heapify(arr: list, n: int, t_idx: int) -> None:
    largest = t_idx
    child_idx = t_idx * 2 + 1

    if child_idx < n and arr[largest] < arr[child_idx]:
        largest = child_idx
    child_idx += 1
    if child_idx < n and arr[largest] < arr[child_idx]:
        largest = child_idx
    if largest != t_idx:
        arr[largest], arr[t_idx] = arr[t_idx], arr[largest]
        heapify(arr, n, largest)


def heap_sort(arr: list, n: int) -> None:
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)
    for i in range(n - 1, -1, -1):
        arr[0], arr[i] = arr[i], arr[0]
        heapify(arr, i, 0)


def print_array(arr: list, size: int) -> None:
    for i in range(0, size):
        print(arr[i], end=' ')
    print()


arr = list(map(int, input().split()))
size = len(arr)

print_array(arr, size)
heap_sort(arr, size)
print_array(arr, size)

'''
Input:
5 4 3 2 1

Output:
5 4 3 2 1
1 2 3 4 5
'''